### ISSUE TITLE
Memory Issues with large number of trees

### ISSUE THREAD

**Issue by ad8437128** [grf-labs/grf#434]
Hello. I have experienced some errors that I think relate to memory issues (perhaps similar to #179) but could be slightly different. I encountered this issue when running `causal_forest` on sample of 30K observations with 100 covariates for 50K trees - *this issue did not occur with fewer than 20K trees*.  I am running R version 3.5.1 and GRF version 0.10.2. 

While it might appear that growing a forest of 50K trees is somewhat excessive, this is not unusual. Section 7 of https://arxiv.org/abs/1610.01271 grows a forest of 100K trees on a sample size of about 330K. 

On Windows PC (64-bit OS, 8GB memory), `causal_forest` throws the following error:
```
Error in instrumental_predict_oob(forest.short, data$default, data$sparse,  : 
  std::bad_alloc
> traceback()
5: stop(list(message = "std::bad_alloc", call = instrumental_predict_oob(forest.short, 
       data$default, data$sparse, num.threads, ci.group.size), cppstack = NULL))
4: instrumental_predict_oob(forest.short, data$default, data$sparse, 
       num.threads, ci.group.size)
3: predict.causal_forest(forest)
2: predict(forest)
1: grf::causal_forest(X = as.matrix(df[covariates]), Y = as.matrix(df["Y"]), 
       W = as.matrix(df["W"]), num.trees = 50000, honesty = TRUE, 
       ci.group.size = 1, tune.parameters = TRUE, seed = cfseed)
```
On a Linux OS on server (with free 250GB memory), the same `causal_forest` routine throws the ff error:
```
 *** caught segfault ***
address 0x7f8587fffff0, cause 'memory not mapped'

Traceback:
 1: instrumental_train(data$default, data$sparse, outcome.index,     treatment.index, instrument.index, as.numeric(tunable.params["mtry"]),     num.trees, num.threads, as.numeric(tunable.params["min.node.size"]),     as.numeric(tunable.params["sample.fraction"]), seed, honesty,     coerce_honesty_fraction(honesty.fraction), ci.group.size,     reduced.form.weight, as.numeric(tunable.params["alpha"]),     as.numeric(tunable.params["imbalance.penalty"]), stabilize.splits,     clusters, samples_per_cluster)
 2: grf::causal_forest(X = as.matrix(df[covariates]), Y = as.matrix(df["Y"]),     W = as.matrix(df["W"]), num.trees = 50000, honesty = TRUE,     ci.group.size = 1, tune.parameters = TRUE, seed = cfseed)
 3: eval(expr, envir, enclos)
 4: eval(expr, envir, enclos)
 5: withVisible(eval(expr, envir, enclos))
 6: withCallingHandlers(withVisible(eval(expr, envir, enclos)), warning = wHandler,     error = eHandler, message = mHandler)
 7: handle(ev <- withCallingHandlers(withVisible(eval(expr, envir,     enclos)), warning = wHandler, error = eHandler, message = mHandler))
 8: timing_fn(handle(ev <- withCallingHandlers(withVisible(eval(expr,     envir, enclos)), warning = wHandler, error = eHandler, message = mHandler)))
 9: evaluate_call(expr, parsed$src[[i]], envir = envir, enclos = enclos,     debug = debug, last = i == length(out), use_try = stop_on_error !=         2L, keep_warning = keep_warning, keep_message = keep_message,     output_handler = output_handler, include_timing = include_timing)
10: evaluate::evaluate(...)
11: evaluate(code, envir = env, new_device = FALSE, keep_warning = !isFALSE(options$warning),     keep_message = !isFALSE(options$message), stop_on_error = if (options$error &&         options$include) 0L else 2L, output_handler = knit_handlers(options$render,         options))
12: in_dir(input_dir(), evaluate(code, envir = env, new_device = FALSE,     keep_warning = !isFALSE(options$warning), keep_message = !isFALSE(options$message),     stop_on_error = if (options$error && options$include) 0L else 2L,     output_handler = knit_handlers(options$render, options)))
13: block_exec(params)
14: call_block(x)
15: process_group.block(group)
16: process_group(group)
17: withCallingHandlers(if (tangle) process_tangle(group) else process_group(group),     error = function(e) {        setwd(wd)        cat(res, sep = "\n", file = output %n% "")        message("Quitting from lines ", paste(current_lines(i),             collapse = "-"), " (", knit_concord$get("infile"),             ") ")    })
18: process_file(text, output)
19: knitr::knit(knit_input, knit_output, envir = envir, quiet = quiet,     encoding = encoding)
20: rmarkdown::render("mwe_segfault_error.Rmd")
An irrecoverable exception occurred. R is aborting now ...
Segmentation fault (core dumped)
```
* Does anyone have any clue about this error, especially regarding why it's related to the number of trees? I do have a sizable memory (at least 250 GB free memory) on the Linux OS on the server.
* Are there any general guidelines on the optimal number of trees beyond "more is better" or "it depends on the application"? In my experience (not reported here), I obtained meaningfully *dissimilar* results (e.g. range of CATE predictions) across different C++ random number generator seeds on the same computer for relatively large causal forests (e.g. 10K - 20K trees). Because such dissimilar results (across seeds) were beyond what one might expect to be *random*, I started experimenting with larger number of trees. 

Thanks!

**Comment by swager** [comment]
@jtibshirani recently made some improvements to GRF memory usage, see #408 and #412. These changes are available in 0.10.3 -- can you try upgrading and see if the problem persists? We're still actively working on this, so any feedback is helpful.

Also, yes, if you're getting meaningfully dissimilar results across runs with different seeds, growing more trees is in general good.

**Comment by ad8437128** [comment]
Thanks @swager and @jtibshirani. After updating to 0.10.3, the error reported above no longer occurred. Also, growing more trees has helped with getting more similar results across different seeds. Thanks for your help.
